subsys/dfu/src/dfu_target.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <zephyr.h>
+#include <logging/log.h>
+#include <dfu/mcuboot.h>
+#include <dfu/dfu_target.h>
+#include "dfu_target_mcuboot.h"
+#include "dfu_target_modem.h"
+
+#ifdef CONFIG_DFU_TARGET_MODEM
+extern struct dfu_target dfu_target_modem;
+#endif
+#ifdef CONFIG_DFU_TARGET_MCUBOOT
+extern struct dfu_target dfu_target_mcuboot;
+#endif
+
+#define MIN_SIZE_IDENTIFY_BUF 32
+
+#define MCUBOOT_IMAGE 1
+#define MODEM_DELTA_IMAGE 2
+
+LOG_MODULE_REGISTER(dfu_target, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+static struct dfu_target *ctx;
+static bool   initialized;
+
+int dfu_target_img_type(const void *const buf, size_t len)
+{
+	if (len < MIN_SIZE_IDENTIFY_BUF) {
+		return -EAGAIN;
+	}
+
+	if (IS_ENABLED(CONFIG_BOOTLOADER_MCUBOOT) &&
+	    dfu_target_mcuboot_identify(buf)) {
+		return MCUBOOT_IMAGE;
+	}
+
+	if (IS_ENABLED(CONFIG_DFU_TARGET_MODEM) &&
+	    dfu_target_modem_identify(buf)) {
+		return MODEM_DELTA_IMAGE;
+	}
+
+	LOG_ERR("No supported image type found");
+	return -ENOTSUP;
+}
+
+int dfu_target_init(int img_type)
+{
+	if (initialized) {
```

Comment:
We kind of duplicate information here - couldn't we simply check if `ctx != NULL`?  No need for `initialized` variable then and double-checks all below.

Reply:
Removed initialized.

Reply:
Looks good. I think its just me being afraid to use NULL for checking anything, but I agree that the new way is much better. 

I will create a PR in to this one with unit tests for stuff like this.